src/TelemetryConsumers/Orleans.TelemetryConsumers.Linux/LinuxEnvironmentStatistics.cs

```diff
@@ -0,0 +1,276 @@
+#define LOG_MEMORY_PERF_COUNTERS
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Orleans.Runtime;
+
+namespace Orleans.Statistics
+{
+    internal class LinuxEnvironmentStatistics : IHostEnvironmentStatistics, ILifecycleParticipant<ISiloLifecycle>, ILifecycleParticipant<IClusterClientLifecycle>, ILifecycleObserver, IDisposable
+    {
+        private readonly ILogger _logger;
+
+        private const float KB = 1024f;
+
+        /// <inheritdoc />
+        public long? TotalPhysicalMemory { get; private set; }
+
+        /// <inheritdoc />
+        public float? CpuUsage { get; private set; }
+
+        /// <inheritdoc />
+        public long? AvailableMemory { get; private set; }
+
+        /// <inheritdoc />
+        public long MemoryUsage => GC.GetTotalMemory(false);
+
+        private readonly TimeSpan MONITOR_PERIOD = TimeSpan.FromSeconds(5);
+
+        private CancellationTokenSource _cts;
+        private Task _monitorTask;
+
+        private static readonly string StartErrMsg_Win = $"Tried to start {nameof(LinuxEnvironmentStatistics)}, but detected OS Windows. " +
+            "Call 'UsePerfCounterEnvironmentStatistics' instead for implementation based on Windows Performance Counters";
+
+        private static readonly string StartErrMsg_Unknown = $"Tried to start {nameof(LinuxEnvironmentStatistics)}, " +
+            $"but detected OS is not Linux: '{RuntimeInformation.OSDescription}'";
+
+        public LinuxEnvironmentStatistics(ILoggerFactory loggerFactory)
+        {
+            _logger = loggerFactory.CreateLogger<LinuxEnvironmentStatistics>();
+        }
+
+        public void Dispose()
+        {
+            _cts?.Dispose();
+            _monitorTask?.Dispose();
+        }
+
+        public void Participate(ISiloLifecycle lifecycle)
+        {
+            lifecycle.Subscribe(ServiceLifecycleStage.RuntimeInitialize, this);
+        }
+
+        public void Participate(IClusterClientLifecycle lifecycle)
+        {
+            lifecycle.Subscribe(ServiceLifecycleStage.RuntimeInitialize, this);
+        }
+
+        public async Task OnStart(CancellationToken ct)
+        {
+            var isLinux = RuntimeInformation.IsOSPlatform(OSPlatform.Linux);
+
+            if (!isLinux)
```

Comment:
I think this check should be in an option validator, this way it will throw when calling `Build()` on the silo builder instead of just displaying a warning at runtime.

You could also check that you have access to the files needed

Reply:
Good point